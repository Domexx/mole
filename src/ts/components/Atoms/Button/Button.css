/* === === === === Button === === === === */

.Button {
  padding: 15px 22px;
  border: 1px solid transparent;
  border-radius: 4px;
  font-family: var(--fnt-primary);
  font-size: 12px;
  font-weight: 500;
  cursor: pointer;
}

.Button--type-default {
  color: var(--clr-white);
  transition: background-color 250ms ease-in-out;
}

.Button--type-default:active,
.Button--type-default:focus {
  outline: 0;
}

.Button--type-slight {
  color: var(--clr-white);
  transition: background-color 250ms ease-in-out;
}

.Button--type-slight:active,
.Button--type-slight:focus {
  outline: 0;
}

.Button--is-Active {
  color: var(--clr-text-primary);
}

.Button--align-left {
  display: table;
  margin-right: auto;
}

.Button--align-center {
  display: table;
  margin-right: auto;
  margin-left: auto;
}

.Button--align-right {
  display: table;
  margin-left: auto;
}
