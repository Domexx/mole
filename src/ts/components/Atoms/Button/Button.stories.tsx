import { Story } from '@storybook/react/types-6-0';
import React, { ComponentProps } from 'react';
import ButtonComponent, { Props } from './Button';

const Component = {
  title: 'Components/Atoms/Button',
  component: ButtonComponent,
  argTypes: {
    type: {
      options: ['default', 'primary', 'error', 'warning'],
      control: { type: 'select' },
    },
    align: {
      options: ['left', 'right', 'center'],
      control: { type: 'inline-radio' },
    },
  },
};

const Template: Story<ComponentProps<typeof ButtonComponent>> = (args) => {
  return <ButtonComponent {...args} />;
};

export const Button = Template.bind({});
export default Component;

// https://storybook.js.org/docs/react/essentials/controls
Button.args = {
  type: 'default',
  align: 'left',
  href: '',
  children: 'Enter',
  disabled: false,
} as Props;
