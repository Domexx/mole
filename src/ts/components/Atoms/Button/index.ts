import loadable from '@loadable/component';

export default loadable(() => {
  return import('./Button');
});
