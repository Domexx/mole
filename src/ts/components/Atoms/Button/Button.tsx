/* eslint-disable react/require-default-props */
import React from 'react';
import { Link } from 'react-router-dom';

import './Button.css';

export type Props = {
  type?: 'default' | 'primary' | 'error' | 'warning' | 'slight';
  align?: 'left' | 'right' | 'center';
  size?: 'normal' | 'narrow';
  href?: string;
  onClick?: (event: React.MouseEvent<HTMLElement>) => void;

  className?: string;
  disabled?: boolean;
};

const Button: React.FC<Props> = ({
  children,
  href,
  onClick,
  disabled,
  className = '',
  align = 'left',
  size = 'normal',
  type = 'default',
}) => {
  let isInternalLink = true;

  if (href) {
    const linkTester = document.createElement('a');
    linkTester.href = href;
    isInternalLink = linkTester.host === location.host;
  }

  const classes = `
    Button
    Button--type-${type}
    Button--size-${size}
    Button--align-${align}
    text-black
    bg-black
    
    ${disabled ? 'Button--disabled' : ''}
    ${className}
  `.trim();

  const globalProps = {
    className: classes,
    ...(disabled && { disabled }),
  };

  if (href && isInternalLink) {
    return (
      <Link {...globalProps} to={href}>
        {children}
      </Link>
    );
  }

  if (href && !isInternalLink) {
    return (
      <a {...globalProps} href={href}>
        {children}
      </a>
    );
  }

  const clickHandler = (event: React.MouseEvent<HTMLElement>) => {
    if (onClick && typeof onClick === 'function') {
      onClick(event);
    }
  };

  return (
    <button {...globalProps} type="button" onClick={clickHandler}>
      {children}
    </button>
  );
};

export default Button;
