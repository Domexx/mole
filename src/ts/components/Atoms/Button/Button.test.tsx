import React from 'react';
import { toJson } from '@test';
import renderer from 'react-test-renderer';
import Button from './Button';

test('Matches Snapshot', () => {
  const component = renderer.create(<Button>Anthony Fu</Button>);
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});
