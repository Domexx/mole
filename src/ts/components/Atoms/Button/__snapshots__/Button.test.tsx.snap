// Vitest Snapshot v1

exports[`Matches Snapshot 1`] = `
<button
  className="Button
    Button--type-default
    Button--size-normal
    Button--align-left
    text-black
    bg-black"
  onClick={[Function]}
  type="button"
>
  Anthony Fu
</button>
`;
